### PR TITLE
add authentication

### PR DESCRIPTION
**Pull request by CUPalex**
Add authentication via Users API

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+     *.        String url (login or logout) and bool isLoggedIn
+     */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        // initialize variables to send them in response
+        String url;
+        Boolean isLoggedIn;
+
+        // get reference to UserService
+        UserService userService = UserServiceFactory.getUserService();
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: When the code is self-explanatory you can drop comments like this. It doesn't add any new information, and therefore redundant.

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        // set url and isLoggedIn vars
+        if (userService.isUserLoggedIn()) {
+            String urlToRedirectToAfterUserLogsOut = "/";
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: Is it used anywhere else? If not, consider removing the variable.

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            url = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+            isLoggedIn = true;
+        } else {
+            String urlToRedirectToAfterUserLogsIn = "/";
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: Same as above.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -196,6 +198,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
      */
     @Override
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        // get reference to UserService
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Redundant

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -196,6 +198,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+        UserService userService = UserServiceFactory.getUserService();
+        // if user is not logged in, s/he can't post comments - send her/him to "/"
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I found 'they', 'them' more useful to refer to a user and easier to read. But it's up to you.

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
 "use strict"
 
-// is called when page loads
+/* is called when page loads */
```
Comment (martemyev) [inline portfolio/src/main/webapp/script.js:17]: Please avoid mixing functional and non-functional (style, etc) changes in one PR. Non-functional changes distract from the functional ones, and make the reviewing process slower. If you'd like to make style changes, you can create a separate PR, and clearly state that it only addresses style and, say, typos in comments - i.e. no functional changes. Such PR is easy to review, and hence fast to merge.

Comment (CUPalex) [reply portfolio/src/main/webapp/script.js:17]: Okay, thanks, will do!
In the next PR there also will be mixed style and functional changes, sorry for that, I just have already made it

**Review by PredaMihaiDragos — APPROVED** [review]
LGTM

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -51,10 +52,52 @@ document.addEventListener("DOMContentLoaded", () => {
...
+            commentsDeleteButton.classList.remove("invisible");
+
+            console.log("LOGGED IN");
+        } else {
+            const loginNav = document.getElementById("login-nav");
+            loginNav.setAttribute("href", url);
+            const loginCommentsLink = document.getElementById("login-comments");
+            loginCommentsLink.setAttribute("href", url);
+
+            console.log("LOGGED OUT");
```
Comment (PredaMihaiDragos) [inline portfolio/src/main/webapp/script.js]: Are these logs for debug or they should stay here?

File: portfolio/src/main/java/com/google/sps/data/AuthSend.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+/* A (login or logout) url + bool isLoggedIn 
+ * object to send in json in AuthServlet.doGet
+ */
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/data/AuthSend.java]: Please try to make comments without spilling implementation details. Something like the following (not sure if I described it 100% correct, but the idea is that I don't mention isLoggedIn, json, and other implementation details that may change in the future)

Class AuthSend represents an object that is used in the authentication servlet
to allow login and logout actions.

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        // initialize variables to send them in response
+        String url;
+        Boolean isLoggedIn;
+
+        UserService userService = UserServiceFactory.getUserService();
+
+        // set url and isLoggedIn vars
+        if (userService.isUserLoggedIn()) {
+            url = userService.createLogoutURL("/");
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java:62]: Same 'url' assignment as on line 65. You can move it out of this 'if' branching to the 'url' intitialization. 'isLoggedIn' in this case also can be initialized without the 'if' structure.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -196,6 +198,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
      */
     @Override
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+        
+        // if users iare not logged in, they can't post comments - send them to "/"
```
Comment (martemyev) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: typo

**Review by martemyev — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java (hunk 9 above)
Comment (CUPalex) [reply portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java:62]: They are not same 'url' assignments -  on line 65 I create login url, and on line 62 - logout. There are different functions.

Comment (martemyev) [reply portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java:62]: Ah, I see, my bad. Then it is fine as it is.

**Review by martemyev — APPROVED** [review]

**Pull request merged**